tree/ntuple/v7/src/RPageStorage.cxx

```diff
@@ -686,20 +699,66 @@ void ROOT::Experimental::Internal::RPagePersistentSink::CommitSealedPage(Descrip
 
 std::vector<ROOT::Experimental::RNTupleLocator>
 ROOT::Experimental::Internal::RPagePersistentSink::CommitSealedPageVImpl(
-   std::span<RPageStorage::RSealedPageGroup> ranges)
+   std::span<RPageStorage::RSealedPageGroup> ranges, const std::vector<bool> &mask)
 {
    std::vector<ROOT::Experimental::RNTupleLocator> locators;
+   locators.reserve(mask.size());
+   std::size_t i = 0;
    for (auto &range : ranges) {
-      for (auto sealedPageIt = range.fFirst; sealedPageIt != range.fLast; ++sealedPageIt)
-         locators.push_back(CommitSealedPageImpl(range.fPhysicalColumnId, *sealedPageIt));
+      for (auto sealedPageIt = range.fFirst; sealedPageIt != range.fLast; ++sealedPageIt) {
+         if (mask[i++])
+            locators.push_back(CommitSealedPageImpl(range.fPhysicalColumnId, *sealedPageIt));
+      }
    }
    return locators;
 }
 
 void ROOT::Experimental::Internal::RPagePersistentSink::CommitSealedPageV(
    std::span<RPageStorage::RSealedPageGroup> ranges)
 {
-   auto locators = CommitSealedPageVImpl(ranges);
+   /// Used in the `originalPages` map
+   struct RSealedPageLink {
+      const RSealedPage *fSealedPage = nullptr; ///< Points to the first occurrence of a page with a specific checksum
+      std::size_t fLocatorIdx = 0;              ///< The index in the locator vector returned by CommitSealedPageVImpl()
+   };
+
+   std::vector<bool> mask;
+   // For every sealed page, stores the corresponding index in the locator vector returned by CommitSealedPageVImpl()
+   std::vector<std::size_t> locatorIndexes;
+   // Maps page checksums to the first sealed page with that checksum
+   std::unordered_map<std::uint64_t, RSealedPageLink> originalPages;
+   std::size_t iLocator = 0;
+   for (auto &range : ranges) {
+      for (auto sealedPageIt = range.fFirst; sealedPageIt != range.fLast; ++sealedPageIt) {
```

Comment:
Maybe worth adding a `reserve()` for both `mask` and `locatorIndexes` here before the inner `for`